snapshots/overlay/overlay.go

```diff
@@ -44,8 +46,19 @@ const upperdirKey = "containerd.io/snapshot/overlay.upperdir"
 type SnapshotterConfig struct {
 	asyncRemove   bool
 	upperdirLabel bool
+	metacopyCtrl  bool
+	metacopyOn    bool
 }
 
+// MetacopyCacheFlag indicates state of the metacopy cache value
+type metacopyCacheFlag int
+
+const (
+	metacopyCacheFalse metacopyCacheFlag = iota
+	metacopyCacheTrue
+	metacopyCacheNone
+)
+
```

Comment:
I don't personally like all of these abstractions. It seems overly complicated, while it can be just simpler.

We need to check if metacopyOn is needed for this image. In that case, if it is needed and supported, we just enable it. If it is needed and not supported, we return an error in `NewSnapshotter()`.

Reply:
Could you, pls, clarify, you mean all "cache" abstraction or exactly these three values (use only two of them)?

Reply:
I think these values are not needed, we probably need a function to read the file, another to write it but we don't need a cache (and this is not really a cache either). Like it is, it adds layers of abstraction that just make it seem way more complicated than it really is: like read a file to see if the snapshot must use metacopyon/off (or the file is not there).

I mentioned in other comments, like [here][], some things that can be simpler. I think the checks that we have to do (like this image needs metadata=on/off or we don't know and is metadata supported in the current kernel) are simple, but reading them with all these abstractions is more complicated.

[here]:https://github.com/containerd/containerd/pull/7141/files#r1156031131

Reply:
I created these values just for simplicity, because there are three use cases -- true/false/nothing, first I thought about simple `*bool`, but it looked less readable than these values. Anyway, I understood your point, I'll try to simplify this.